build: support OpenBSD in depends

fanquake opened the pull request:
Completes the BSD trifecta. No Qt. Includes a commit that adds a default build_TAR to depends, to make it easier to override and use gnu/gtar, as OpenBSDs tar does not support some of the options we use. Also includes this patch, https://github.com/boostorg/test/commit/684f067dde3b798877655cdda4eab8d7c26b2510, to fix a compilation issue in Boost test.

Have tested on OpenBSD 7.0 with: 
```bash
gmake -C depends -j9
./autogen.sh
CONFIG_SITE=/home/vagrant/bitcoin/depends/amd64-unknown-openbsd7.0/share/config.site ./configure --disable-external-signer MAKE=gmake
gmake check -j9
```

fanquake committed:
build: add support for OpenBSD to depends

fanquake committed:
build: add a default build tar in depends

This is so we can override it later for BSDs.

MarcoFalke commented on depends/packages/libevent.mk:
```diff
@@ -12,6 +12,7 @@ define $(package)_set_vars
   $(package)_config_opts += --disable-dependency-tracking --enable-option-checking
   $(package)_config_opts_release=--disable-debug-mode
   $(package)_config_opts_linux=--with-pic
+  $(package)_config_opts_openbsd=--with-pic  
```
```suggestion
  $(package)_config_opts_openbsd=--with-pic
```

fanquake replied:
Addressed.

theStack commented:
Concept ACK

theStack reviewed (commented):
I could successfully build and run the tests on OpenBSD 7.0 with the instructions provided in the PR description. However, I had to prepend `TAR=gtar` to the first gmake call, otherwise it would fail due to the base system `tar` not supporting the needed options.

Is this supposed to be specified by the user or was the idea to set `TAR=gtar` by default in the Makefile on OpenBSD?

laanwj commented:
We need to document the needed packages/ports somewhere (in `build-openbsd.md` would be the most likely place? or maybe in depend's `README.md`):

```
pkg_add gtar bash
```

- `gtar` as mentioned
- `bash` for `gen_id` only i think, not installing it will give the following output but it seems to work otherwise:
```
env: ./gen_id: No such file or directory
env: ./gen_id: No such file or directory
```

fanquake commented:
Rebased.

> Is this supposed to be specified by the user or was the idea to set TAR=gtar by default in the Makefile on OpenBSD?

Made a (currently untested) change such that using `gtar` should be set automatically, so no need to specify `TAR=gtar` when building.

> We need to document the needed packages/ports somewhere

I've added these to the depends README.

laanwj commented:
Thanks! Updated to commit 8793ee3a9df92f9f37c2abc5be3e088b56751e69.
I don't think the automatic workaround is working. `gtar` is installed but:
```
Extracting native_b2...
(SHA256) /home/user/bitcoin/depends/sources/boost_1_71_0.tar.bz2: OK
tar: unknown option -- -
usage: tar {crtux}[014578befHhjLmNOoPpqsvwXZz]
           [blocking-factor | archive | replstr] [-C directory] [-I file]
           [file ...]
       tar {-crtux} [-014578eHhjLmNOoPpqvwXZz] [-b blocking-factor]
           [-C directory] [-f archive] [-I file] [-s replstr] [file ...]
gmake: *** [funcs.mk:282: /home/user/bitcoin/depends/work/build/amd64-unknown-openbsd7.0/native_b2/1_71_0-091b88c6ae3/.stamp_extracted] Error 1
```
Adding `TAR=gtar` gets it through that.

laanwj commented:
zeromq build fails with:
```
Provide an AUTOCONF_VERSION environment variable, please
autogen.sh: error: autoreconf exited with status 127
gmake: *** [funcs.mk:283: /home/user/bitcoin/depends/work/build/amd64-unknown-openbsd7.0/zeromq/4.3.4-a71c7ab8497/./.stamp_configured] Error 1
```
same as or the normal build instructions, i suppose:
```
export AUTOCONF_VERSION=2.69
export AUTOMAKE_VERSION=1.16
```

fanquake commented:
> I don't think the automatic workaround is working. gtar is installed but:

Sorry, should be addressed now. I've reworked how it's set. Will test shortly myself.

laanwj commented:
The gtar issue is fixed. I didn't have to explicitly specify it this time. It succeeds in building the default dependencies (except Qt, as documented).

However, running into the following during configure:
```
configure:29529: checking for Berkeley DB C++ headers
configure:29559: c++ -std=c++17 -c -g -O2  -DHAVE_BUILD_INFO -DPROVIDE_FUZZ_MAIN_FUNCTION conftest.cpp >&5
conftest.cpp:73:18: fatal error: 'bdb4.8/db_cxx.h' file not found
        #include <bdb4.8/db_cxx.h>
                 ^~~~~~~~~~~~~~~~~
```
It looks like the header is installed but in a different path than expected:
```
$ find . -name db_cxx.h
./depends/amd64-unknown-openbsd7.0/include/db_cxx.h
```

laanwj commented:
Tested ACK 471d15536f12889077578884fca33d1b0930fd66

theStack reviewed (approved):
Tested ACK 471d15536f12889077578884fca33d1b0930fd66
(on OpenBSD 7.0, amd64)

`gtar` is now selected automatically. If it's missing, the message "gtar: not found" appears and the build is aborted:
```
$ gmake -C depends/ -j4
gmake: Entering directory '/home/honey/bitcoin_prrev/depends'
Extracting native_b2...
(SHA256) /home/honey/bitcoin_prrev/depends/sources/boost_1_71_0.tar.bz2: OK
/bin/sh: gtar: not found
gmake: *** [funcs.mk:282: /home/honey/bitcoin_prrev/depends/work/build/amd64-unknown-openbsd7.0/native_b2/1_71_0-c97b4ba1fa7/.stamp_extracted] Error 127
gmake: Leaving directory '/home/honey/bitcoin_prrev/depends'
```
After installing `gtar` (`$ doas pkg_add gtar`) the build succeeds with the instruction in the PR description.
> I did a depends build on OpenBSD 7.0, then built bitcoind, and ran both the unit tests and functional tests succesfully. I also checked with ldd that all depends were linked statically.

Did the same and can confirm that everything succeeds and the only dynamically linked libraries in `bitcoind` belong to the base system (libpthread, libz, libm, libc++, libc++abi, libc, ld) and not to depends.
